make entire option rows clickable in settings page

dreacot opened the pull request:
Resolves https://github.com/planetdecred/godcr/issues/310

This PR makes the entire option rows in the settings page clickable to enable the users to easily use/access the option

dreacot committed:
make entire option rows clickable in settings page

oshorefueled commented on go.mod:
```diff
@@ -10,6 +10,7 @@ require (
 	github.com/decred/slog v1.1.0
 	github.com/jessevdk/go-flags v1.4.1-0.20200711081900-c17162fe8fd7
 	github.com/jrick/logrotate v1.0.0
+	github.com/markbates/pkger v0.17.1
```
pkger doesn't seem to be used in this PR

dreacot committed:
increase the clickable area for the option rows with navigation icon

fix merge conflicts

oshorefueled commented on ui/settings_page.go:
```diff
@@ -12,13 +12,18 @@ import (
 
 const PageSettings = "Settings"
 
+type optionRowHandler struct {
```
This isn't a handler. It should probably be just "optionRow"

oshorefueled commented on go.sum:
```diff
@@ -1,25 +1,35 @@
 decred.org/dcrwallet v1.6.0-rc4 h1:5IT6mFa+2YMqenu6aE2LetD0N8QSUVFyAFl205PvIIE=
 decred.org/dcrwallet v1.6.0-rc4/go.mod h1:lsrNbuKxkPGeHXPufxNTckwQopCEDz0r3t0a8JCKAmU=
 dmitri.shuralyov.com/gpu/mtl v0.0.0-20190408044501-666a987793e9/go.mod h1:H6x//7gZCb22OMCxBHrMx7a5I7Hp++hsVxbQ4BYO7hU=
+gioui.org v0.0.0-20200722191435-e381ff40d66b h1:gnY/k6FSe0jYKiVFDG0ZSnBRry1zgh5c841NFifuQTA=
```
The PR shouldn't make changes to the go.mod or go.sum files.

oshorefueled commented on ui/settings_page.go:
```diff
@@ -252,7 +276,7 @@ func (pg *settingsPage) agent() layout.Widget {
 	return layout.Inset{Bottom: values.MarginPadding10}.Layout(gtx, func(gtx C) D {
 		return pg.theme.Card().Layout(gtx, func(gtx C) D {
-			return layout.UniformInset(values.MarginPadding15).Layout(gtx, func(gtx C) D {
+			return layout.Inset{Top: values.MarginPadding15, Left: values.MarginPadding15, Right: values.MarginPadding15}.Layout(gtx, func(gtx C) D {
```
MarginPadding15 can be assigned to a variable to break the code to the next line.

dreacot replied:
i'm not sure what you mean here?

oshorefueled commented on ui/settings_page.go:
```diff
@@ -123,15 +132,28 @@ func (pg *settingsPage) general() layout.Widget {
+						return layout.Stack{}.Layout(gtx,
+							layout.Stacked(func(gtx C) D {
+								list := layout.List{Axis: layout.Vertical}
+								return list.Layout(gtx, 1, func(gtx C, i int) D {
```
I'm not sure why a list layout is needed here.

dreacot committed:
rename optionRowHandler to optionRow
make user agent and connect to peer rows clickable
add margin to a variable, to enable code break

oshorefueled commented on ui/settings_page.go:
```diff
@@ -12,15 +12,20 @@ import (
 
+type optionRow struct {
+	clickable *widget.Clickable
+	icon      decredmaterial.IconButton
```
Having an clickable widget and an icon button do the same thing seems redundant. The clickable widget and an icon might just be all that's needed. Reassess the code and see if you really need to create a new struct type for this.

oshorefueled commented on ui/settings_page.go:
```diff
@@ -123,17 +140,21 @@ func (pg *settingsPage) general() layout.Widget {
+									}),
+								)
+							})
+						})
```
This code block seems to be reused for a number of option rows, you can consider making it a function or method and reusing that.

oshorefueled commented on ui/settings_page.go:
```diff
@@ -404,7 +438,7 @@ func (pg *settingsPage) handle(common pageCommon) {
 
 	userAgentKey := dcrlibwallet.UserAgentConfigKey
-	for pg.updateUserAgent.Button.Clicked() {
+	for pg.updateUserAgent.clickable.Clicked() || pg.updateUserAgent.icon.Button.Clicked() {
```
Removing the icon button and using just an icon would remove the need for this or condition.

dreacot committed:
change icon button to just an icon
make function to handle clickable rows to prevent redundant code

oshorefueled commented on ui/settings_page.go:
```diff
@@ -123,17 +128,21 @@ func (pg *settingsPage) general() layout.Widget {
-						)
+					return decredmaterial.Clickable(gtx, pg.currencyConversion, func(gtx C) D {
+						return layout.Inset{Top: values.MarginPadding5, Bottom: values.MarginPaddingMinus5}.Layout(gtx, func(gtx C) D {
+							return pg.subSection(gtx, "Currency conversion", func(gtx C) D {
```
why wasn't the clickableRow method used for currency conversion. It looks like it would work here.

dreacot replied:
Yes it would, but the functionality hasn't been implemented yet, once the functionality is implemented, we can then pass in the required label as the parameter

Else I'd have to put an if else statement in the clickableRow function to accommodate the currency conversion

oshorefueled commented on ui/settings_page.go:
```diff
@@ -283,6 +280,23 @@ func (pg *settingsPage) subSectionSwitch(gtx layout.Context, title string, optio
 	})
 }
 
+func (pg *settingsPage) clickableRow(gtx layout.Context, title string, clickable *widget.Clickable, icon *widget.Icon, label decredmaterial.Label) layout.Dimensions {
```
If the arguments you'll be passing to a method is becoming too many, you can create a struct type and pass  that instead. For instance, for clickableRow, you can define the method to accept layout.Context and a type "row" as arguments. The type "row" would contain the remaining the parameters you defined on the method.

oshorefueled commented on ui/settings_page.go:
```diff
@@ -283,6 +280,23 @@ func (pg *settingsPage) subSectionSwitch(gtx layout.Context, title string, optio
+	})
+}
+
 func (pg *settingsPage) conditionalDisplay(gtx layout.Context, body layout.Widget) layout.Dimensions {
```
Kindly add a boolean as one of the parameters of this method so it can accept the condition as an argument. With the "if" statement  inside the method.

dreacot replied:
I'm not sure I fully get what you mean here

oshorefueled replied:
something like this 
`
func (pg *settingsPage) conditionalDisplay(gtx layout.Context, display bool, body layout.Widget) layout.Dimensions { 
  if display { 
    return layout.Flex{Axis: layout.Vertical}.Layout(gtx, 
      layout.Rigid(
       pg.lineSeparator()), 
       layout.Rigid(body), 
      ) 
  } 
  return layout.Dimensions{}
}
`

oshorefueled commented on ui/settings_page.go:
```diff
@@ -54,14 +60,15 @@ func (win *Window) SettingsPage(common pageCommon) layout.Widget {
-		changeStartupPass:   common.theme.PlainIconButton(new(widget.Clickable), icon),
+		currencyConversion:  currencyConversion,
+		updateConnectToPeer: updateConnectToPeer,
+		updateUserAgent:     updateUserAgent,
```
You don't need to create variables for the Clickable types, you can just use new(widget.Cickable) here.

dreacot committed:
remove unnecessary variables
add a  bool parameter, to the conditionalDisplay function
create row struct as an argument for the clickableRow function
